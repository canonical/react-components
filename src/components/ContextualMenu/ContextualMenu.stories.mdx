import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import ContextualMenu from "./ContextualMenu";

<Meta
  title="ContextualMenu"
  component={ContextualMenu}
  argTypes={{
    children: {
      control: {
        disable: true,
      },
    },
    links: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => (
  <div className="u-align--center">
    <ContextualMenu {...args} />
  </div>
);

### Contextual Menu

This is a [React](https://reactjs.org/) component for the Vanilla [Contextual menu](https://docs.vanillaframework.io/patterns//contextual-menu/).

A contextual menu can be used in conjunction with any page element to provide a contextual menu.

### Props

<ArgsTable of={ContextualMenu} />

### Default

The contextual menu will provide a visual wrapper to any provided children. Visibility can be toggled via the `visible` prop.

<Canvas>
  <Story
    name="Default"
    args={{
      children: <span style={{ padding: "1rem" }}>This is a menu.</span>,
      closeOnOutsideClick: false,
      constrainPanelWidth: false,
      position: "left",
      visible: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Toggle

<Canvas>
  <Story
    name="Toggle"
    args={{
      links: [
        {
          children: "Link 1",
          onClick: () => {},
        },
        {
          children: "Link 2",
          onClick: () => {},
        },
      ],
      hasToggleIcon: true,
      position: "right",
      toggleLabel: "Click me!",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Child function

<Canvas>
  <Story
    name="Child function"
    args={{
      children: (close) => <Button onClick={close}>child function</Button>,
      hasToggleIcon: true,
      position: "right",
      toggleLabel: "Click me!",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Child element

<Canvas>
  <Story
    name="Child element"
    args={{
      children: <Button>child element</Button>,
      hasToggleIcon: true,
      position: "right",
      toggleLabel: "Click me!",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
