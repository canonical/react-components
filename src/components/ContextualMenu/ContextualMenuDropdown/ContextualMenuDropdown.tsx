import classNames from "classnames";
import React, { useCallback, useEffect, useRef, useState } from "react";
import type { HTMLProps, ReactNode } from "react";

import { useWindowFitment } from "hooks";
import Button from "../../Button";
import type { ButtonProps } from "../../Button";
import type { WindowFitment } from "hooks";

export enum Label {
  Dropdown = "submenu",
}

/**
 * The type of the menu links.
 * @template L - The type of the link props.
 */
export type MenuLink<L = null> = string | ButtonProps<L> | ButtonProps<L>[];

export type Position = "left" | "center" | "right";

/**
 * The props for the ContextualMenuDropdown component.
 * @template L - The type of the link props.
 */
export type Props<L = null> = {
  adjustedPosition?: Position;
  autoAdjust?: boolean;
  closePortal?: (evt?: MouseEvent) => void;
  constrainPanelWidth?: boolean;
  dropdownClassName?: string;
  dropdownContent?: ReactNode;
  id?: string;
  isOpen?: boolean;
  links?: MenuLink<L>[];
  position?: Position;
  positionCoords?: ClientRect;
  positionNode?: HTMLElement;
  setAdjustedPosition?: (position: Position) => void;
  wrapperClass?: string;
} & HTMLProps<HTMLSpanElement>;

/**
 * Calculate the styles for the menu.
 * @param position - The menu position.
 * @param positionCoords - The coordinates of the position node.
 * @param constrainPanelWidth - Whether the menu width should be constrained to the position width.
 */
const getPositionStyle = (
  position: Position,
  positionCoords: Props["positionCoords"],
  constrainPanelWidth: Props["constrainPanelWidth"]
) => {
  if (!positionCoords) {
    return null;
  }
  const { height, left, top, width } = positionCoords;
  const topPos = top + height + (window.scrollY || 0);
  let leftPos = left;

  switch (position) {
    case "left":
      leftPos = left;
      break;
    case "center":
      leftPos = left + width / 2;
      break;
    case "right":
      leftPos = left + width;
      break;
    default:
      break;
  }

  return {
    position: "absolute",
    left: leftPos,
    top: topPos,
    // The width only needs to be set if the width is to be constrained.
    ...(constrainPanelWidth ? { width } : null),
  };
};

/**
 * Calculate the adjusted position in relation to the window.
 * @param position - The requested position.
 * @param fitsWindow - The window fitment info.
 * @return The new position.
 */
export const adjustForWindow = (
  position: Position,
  fitsWindow: WindowFitment
): Position => {
  let newPosition: string = position;
  if (!fitsWindow.fromRight.fitsLeft && newPosition === "right") {
    newPosition = "left";
  }
  if (!fitsWindow.fromLeft.fitsRight && newPosition === "left") {
    newPosition = "right";
  }
  // If the menu doesn't fit to the left or the right then center it.
  if (
    !fitsWindow.fromLeft.fitsRight &&
    !fitsWindow.fromRight.fitsLeft &&
    (newPosition === "left" || newPosition === "right")
  ) {
    newPosition = "center";
  }
  // If the menu doesn't fit when centered then find a new position.
  if (
    newPosition === "center" &&
    (!fitsWindow.fromCenter.fitsCentered.fitsRight ||
      !fitsWindow.fromCenter.fitsCentered.fitsLeft)
  ) {
    if (fitsWindow.fromLeft.fitsRight) {
      newPosition = "left";
    }
    if (fitsWindow.fromRight.fitsLeft) {
      newPosition = "right";
    }
  }
  return newPosition as Position;
};

/**
 * Generate a menu link
 * @template L - The type of the link props.
 * @param link - A button
 * @param key - A key for the DOM.
 * @param closePortal - The function to close the portal.
 */
const generateLink = <L,>(
  link: ButtonProps,
  key: React.Key,
  closePortal: Props["closePortal"]
) => {
  const { children, className, onClick, ...props } = link;
  return (
    <Button<L>
      className={classNames("p-contextual-menu__link", className)}
      key={key}
      onClick={
        onClick
          ? (evt) => {
              closePortal(evt.nativeEvent);
              onClick(evt);
            }
          : null
      }
      {...props}
    >
      {children}
    </Button>
  );
};

const ContextualMenuDropdown = <L,>({
  adjustedPosition,
  autoAdjust,
  closePortal,
  constrainPanelWidth,
  dropdownClassName,
  dropdownContent,
  id,
  isOpen,
  links,
  position,
  positionCoords,
  positionNode,
  setAdjustedPosition,
  wrapperClass,
  ...props
}: Props<L>): JSX.Element => {
  const dropdown = useRef();
  const [positionStyle, setPositionStyle] = useState(
    getPositionStyle(adjustedPosition, positionCoords, constrainPanelWidth)
  );

  // Update the styles to position the menu.
  const updatePositionStyle = useCallback(() => {
    setPositionStyle(
      getPositionStyle(adjustedPosition, positionCoords, constrainPanelWidth)
    );
  }, [adjustedPosition, positionCoords, constrainPanelWidth]);

  // Update the position when the window fitment info changes.
  const onUpdateWindowFitment = useCallback(
    (fitsWindow) => {
      setAdjustedPosition(adjustForWindow(position, fitsWindow));
    },
    [position, setAdjustedPosition]
  );

  // Handle adjusting the position of the dropdown so that it remains on screen.
  useWindowFitment(
    dropdown.current,
    positionNode,
    onUpdateWindowFitment,
    0,
    isOpen && autoAdjust
  );

  // Update the styles when the position changes.
  useEffect(() => {
    updatePositionStyle();
  }, [adjustedPosition, updatePositionStyle]);

  return (
    <span
      className={wrapperClass}
      style={positionStyle as React.CSSProperties}
      {...props}
    >
      <span
        className={classNames("p-contextual-menu__dropdown", dropdownClassName)}
        id={id}
        aria-hidden={isOpen ? "false" : "true"}
        aria-label={Label.Dropdown}
        ref={dropdown}
        style={
          constrainPanelWidth && positionStyle?.width
            ? { width: positionStyle.width, minWidth: 0, maxWidth: "none" }
            : null
        }
      >
        {dropdownContent
          ? dropdownContent
          : links.map((item, i) => {
              if (Array.isArray(item)) {
                return (
                  <span className="p-contextual-menu__group" key={i}>
                    {item.map((link, j) =>
                      generateLink<L>(link, j, closePortal)
                    )}
                  </span>
                );
              } else if (typeof item === "string") {
                return (
                  <div className="p-contextual-menu__non-interactive" key={i}>
                    {item}
                  </div>
                );
              }
              return generateLink<L>(item, i, closePortal);
            })}
      </span>
    </span>
  );
};

export default ContextualMenuDropdown;
