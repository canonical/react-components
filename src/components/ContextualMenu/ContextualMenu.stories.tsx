import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Button from "../Button";
import ContextualMenu from "./ContextualMenu";

const ScrollTemplate = (args) => (
  <div
    style={{
      maxWidth: "30rem",
      maxHeight: "10rem",
      overflow: "auto",
      padding: "1rem",
    }}
  >
    <ContextualMenu {...args} />
    {Array(3).fill(
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Consequuntur
        cum dicta beatae nostrum eligendi similique earum, dolorem fuga quis,
        sequi voluptates architecto ipsa dolorum eaque rem expedita inventore
        voluptas odit aspernatur alias molestias facere.
      </p>
    )}
    <ContextualMenu {...args} />
  </div>
);

const Template = (args) => (
  <div className="u-align--center">
    <ContextualMenu {...args} />
  </div>
);

const meta: Meta<typeof ContextualMenu> = {
  component: ContextualMenu,
  render: Template,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        disable: true,
      },
    },

    links: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof ContextualMenu>;

/**
 * The contextual menu will provide a visual wrapper to any provided children. Visibility can be toggled via the `visible` prop.
 */
export const Default: Story = {
  name: "Default",

  args: {
    children: (
      <span
        style={{
          padding: "1rem",
        }}
      >
        This is a menu.
      </span>
    ),

    closeOnOutsideClick: false,
    constrainPanelWidth: false,
    position: "left",
    visible: true,
  },
};

export const Toggle: Story = {
  name: "Toggle",

  args: {
    links: [
      {
        children: "Link 1",
        onClick: () => {},
      },
      {
        children: "Link 2",
        onClick: () => {},
      },
    ],

    hasToggleIcon: true,
    position: "right",
    toggleLabel: "Click me!",
  },
};

export const OverflowingContainer: Story = {
  render: ScrollTemplate.bind({}),
  name: "Overflowing container",

  args: {
    links: [
      {
        children: "Link 1",
        onClick: () => {},
      },
      {
        children: "Long Link 2",
        onClick: () => {},
      },
    ],

    hasToggleIcon: true,
    position: "right",
    toggleLabel: "Click me!",
  },
};

export const ChildFunction: Story = {
  name: "Child function",

  args: {
    children: (close) => <Button onClick={close}>child function</Button>,
    hasToggleIcon: true,
    position: "right",
    toggleLabel: "Click me!",
  },
};

export const ChildElement: Story = {
  name: "Child element",

  args: {
    children: <Button>child element</Button>,
    hasToggleIcon: true,
    position: "right",
    toggleLabel: "Click me!",
  },
};
