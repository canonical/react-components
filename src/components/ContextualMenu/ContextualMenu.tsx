import classNames from "classnames";
import React, { useCallback, useEffect, useId, useRef, useState } from "react";
import type { HTMLProps } from "react";
import usePortal from "react-useportal";

import { useListener, usePrevious } from "hooks";
import Button from "../Button";
import type { ButtonProps } from "../Button";
import ContextualMenuDropdown from "./ContextualMenuDropdown";
import type { ContextualMenuDropdownProps } from "./ContextualMenuDropdown";
import type { MenuLink, Position } from "./ContextualMenuDropdown";
import { ClassName, PropsWithSpread, SubComponentProps } from "types";

export enum Label {
  Toggle = "Toggle menu",
}

/**
 * The props for the ContextualMenu component.
 * @template L - The type of the link props.
 */
export type Props<L> = PropsWithSpread<
  {
    /**
     * Whether the menu should adjust its horizontal position to fit on the screen.
     */
    autoAdjust?: boolean;
    /**
     * The menu content (if the links prop is not supplied).
     */
    children?: ContextualMenuDropdownProps["dropdownContent"];
    /**
     * An optional class to apply to the wrapping element.
     */
    className?: ClassName;
    /**
     * Whether the menu should close when the escape key is pressed.
     */
    closeOnEsc?: boolean;
    /**
     * Whether the menu should close when clicking outside the menu.
     */
    closeOnOutsideClick?: boolean;
    /**
     * Whether the menu's width should match the toggle's width.
     */
    constrainPanelWidth?: boolean;
    /**
     * An optional class to apply to the dropdown.
     */
    dropdownClassName?: string | null;
    /**
     * Additional props to pass to the dropdown.
     */
    dropdownProps?: SubComponentProps<ContextualMenuDropdownProps>;
    /**
     * Whether the toggle should display a chevron icon.
     */
    hasToggleIcon?: boolean;
    /**
     * A list of links to display in the menu (if the children prop is not supplied.)
     */
    links?: MenuLink<L>[] | null;
    /**
     * A function to call when the menu is toggled.
     */
    onToggleMenu?: (isOpen: boolean) => void | null;
    /**
     * The horizontal position of the menu.
     */
    position?: Position | null;
    /**
     * An element to make the menu relative to.
     */
    positionNode?: HTMLElement | null;
    /**
     * Whether the dropdown should scroll if it is too long to fit on the screen.
     */
    scrollOverflow?: boolean;
    /**
     * The appearance of the toggle button.
     */
    toggleAppearance?: ButtonProps["appearance"] | null;
    /**
     * A class to apply to the toggle button.
     */
    toggleClassName?: string | null;
    /**
     * Whether the toggle button should be disabled.
     */
    toggleDisabled?: boolean;
    /**
     * The toggle button's label.
     */
    toggleLabel?: React.ReactNode | null;
    /**
     * Whether the toggle lable or icon should appear first.
     */
    toggleLabelFirst?: boolean;
    /**
     * Additional props to pass to the toggle button.
     */
    toggleProps?: SubComponentProps<ButtonProps>;
    /**
     * Whether the menu should be visible.
     */
    visible?: boolean;
  },
  HTMLProps<HTMLSpanElement>
>;

/**
 * Get the node to use for positioning the menu.
 * @param wrapper - The component's wrapping element.
 * @param positionNode - A positioning node, if supplied.
 * @return A node or null.
 */
const getPositionNode = (
  wrapper: HTMLElement,
  positionNode?: HTMLElement
): HTMLElement | null => {
  if (positionNode) {
    return positionNode;
  } else if (wrapper) {
    // We want to position the menu in relation to the toggle, if it exists.
    const toggle = wrapper.querySelector<HTMLElement>(
      ".p-contextual-menu__toggle"
    );
    return toggle || wrapper;
  }
  return null;
};

/**
 * Whether the positioning node is visible.
 * @param positionNode - The node that is used to position the menu.
 * @return Whether the positioning node is visible.
 */
const getPositionNodeVisible = (positionNode: HTMLElement) => {
  return !positionNode || positionNode.offsetParent !== null;
};

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Contextual menu](https://docs.vanillaframework.io/patterns//contextual-menu/).
 *
 * A contextual menu can be used in conjunction with any page element to provide a contextual menu.
 */
const ContextualMenu = <L,>({
  autoAdjust = true,
  children,
  className,
  closeOnEsc = true,
  closeOnOutsideClick = true,
  constrainPanelWidth,
  dropdownClassName,
  dropdownProps,
  hasToggleIcon,
  links,
  onToggleMenu,
  position = "right",
  positionNode,
  scrollOverflow,
  toggleAppearance,
  toggleClassName,
  toggleDisabled,
  toggleLabel,
  toggleLabelFirst = true,
  toggleProps,
  visible = false,
  ...wrapperProps
}: Props<L>): JSX.Element => {
  const id = useId();
  const wrapper = useRef<HTMLDivElement | null>(null);
  const [positionCoords, setPositionCoords] = useState<DOMRect>();
  const [adjustedPosition, setAdjustedPosition] = useState(position);
  const hasToggle = hasToggleIcon || toggleLabel;

  useEffect(() => {
    setAdjustedPosition(position);
  }, [position, autoAdjust]);

  // Update the coordinates of the position node.
  const updatePositionCoords = useCallback(() => {
    const parent = getPositionNode(wrapper.current, positionNode);
    if (!parent) {
      return null;
    }
    setPositionCoords(parent.getBoundingClientRect());
  }, [wrapper, positionNode]);

  const { openPortal, closePortal, isOpen, Portal } = usePortal({
    closeOnEsc,
    closeOnOutsideClick,
    isOpen: visible,
    onOpen: () => {
      // Call the toggle callback, if supplied.
      onToggleMenu && onToggleMenu(true);
      // When the menu opens then update the coordinates of the parent.
      updatePositionCoords();
    },
    onClose: () => {
      // Call the toggle callback, if supplied.
      onToggleMenu && onToggleMenu(false);
    },
  });

  const previousVisible = usePrevious(visible);
  const labelNode =
    toggleLabel && typeof toggleLabel === "string" ? (
      <span>{toggleLabel}</span>
    ) : React.isValidElement(toggleLabel) ? (
      toggleLabel
    ) : null;
  const contextualMenuClassName = classNames(className, "p-contextual-menu", {
    [`p-contextual-menu--${adjustedPosition}`]: adjustedPosition !== "right",
  });

  // Update the coordinates of the wrapper once it mounts to the dom. This uses
  // The callback ref pattern:
  // https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node
  const wrapperRef = useCallback(
    (node) => {
      wrapper.current = node;
      if (node !== null) {
        updatePositionCoords();
      }
    },
    [updatePositionCoords]
  );

  // Handle controlling updates to the menu visibility from outside
  // the component.
  useEffect(() => {
    if (visible !== previousVisible) {
      if (visible && !isOpen) {
        openPortal();
      } else if (!visible && isOpen) {
        closePortal();
      }
    }
  }, [closePortal, openPortal, visible, isOpen, previousVisible]);

  const onResize = useCallback(
    (evt) => {
      const parent = getPositionNode(wrapper.current, positionNode);
      if (parent && !getPositionNodeVisible(parent)) {
        // Hide the menu if the item has become hidden. This might happen in
        // a responsive table when columns become hidden as the page
        // becomes smaller.
        closePortal(evt);
      } else {
        // Update the coordinates so that the menu stays relative to the
        // toggle button.
        updatePositionCoords();
      }
    },
    [closePortal, positionNode, updatePositionCoords]
  );

  const onScroll = useCallback(
    (e) => {
      const parent = getPositionNode(wrapper.current, positionNode);
      // update position if the scroll event is triggered by the parent of the menu
      if (parent && e.target.contains(parent)) {
        // Update the coordinates so that the menu stays relative to the
        // toggle button.
        updatePositionCoords();
      }
    },
    [positionNode, updatePositionCoords]
  );

  useListener(window, onResize, "resize", true, isOpen);
  useListener(window, onScroll, "scroll", false, isOpen, true);

  return (
    <span
      className={contextualMenuClassName}
      ref={wrapperRef}
      {...wrapperProps}
    >
      {hasToggle ? (
        <Button
          appearance={toggleAppearance}
          aria-controls={id}
          aria-expanded={isOpen ? "true" : "false"}
          aria-label={toggleLabel ? null : Label.Toggle}
          aria-pressed={isOpen ? "true" : "false"}
          aria-haspopup="true"
          className={classNames("p-contextual-menu__toggle", toggleClassName)}
          disabled={toggleDisabled}
          hasIcon={hasToggleIcon}
          onClick={(evt: React.MouseEvent) => {
            if (!isOpen) {
              openPortal(evt);
            } else {
              closePortal(evt);
            }
          }}
          type="button"
          {...toggleProps}
        >
          {toggleLabelFirst ? labelNode : null}
          {hasToggleIcon ? (
            <i
              className={classNames(
                "p-icon--chevron-down p-contextual-menu__indicator",
                {
                  "is-light": ["negative", "positive"].includes(
                    toggleAppearance
                  ),
                }
              )}
            ></i>
          ) : null}
          {toggleLabelFirst ? null : labelNode}
        </Button>
      ) : null}
      {isOpen && (
        <Portal>
          <ContextualMenuDropdown<L>
            adjustedPosition={adjustedPosition}
            autoAdjust={autoAdjust}
            handleClose={closePortal}
            constrainPanelWidth={constrainPanelWidth}
            dropdownClassName={dropdownClassName}
            dropdownContent={children}
            id={id}
            isOpen={isOpen}
            links={links}
            position={position}
            positionCoords={positionCoords}
            contextualMenuClassName={contextualMenuClassName}
            positionNode={getPositionNode(wrapper.current)}
            scrollOverflow={scrollOverflow}
            setAdjustedPosition={setAdjustedPosition}
            {...dropdownProps}
          />
        </Portal>
      )}
    </span>
  );
};

export default ContextualMenu;
