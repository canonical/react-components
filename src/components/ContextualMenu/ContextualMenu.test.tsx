import { render, screen, within } from "@testing-library/react";
import React from "react";

import ContextualMenu, { Label } from "./ContextualMenu";
import { Label as DropdownLabel } from "./ContextualMenuDropdown/ContextualMenuDropdown";
import userEvent from "@testing-library/user-event";

describe("ContextualMenu ", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders", () => {
    render(<ContextualMenu data-testid="menu" links={[]} />);
    expect(screen.getByTestId("menu")).toMatchSnapshot();
  });

  it("can be aligned to the right", () => {
    render(<ContextualMenu data-testid="menu" links={[]} position="right" />);
    const menu = screen.getByTestId("menu");
    expect(menu).toHaveClass("p-contextual-menu");
    expect(menu).not.toHaveClass("p-contextual-menu--left");
    expect(menu).not.toHaveClass("p-contextual-menu--center");
  });

  it("can be aligned to the left", () => {
    render(<ContextualMenu data-testid="menu" links={[]} position="left" />);
    expect(screen.getByTestId("menu")).toHaveClass("p-contextual-menu--left");
  });

  it("can be aligned to the center", () => {
    render(<ContextualMenu data-testid="menu" links={[]} position="center" />);
    expect(screen.getByTestId("menu")).toHaveClass("p-contextual-menu--center");
  });

  it("can have a toggle button", () => {
    render(<ContextualMenu links={[]} toggleLabel="Toggle" />);
    expect(screen.getByRole("button", { name: "Toggle" })).toBeInTheDocument();
  });

  it("can be disabled", () => {
    render(<ContextualMenu links={[]} toggleDisabled toggleLabel="Toggle" />);
    expect(screen.getByRole("button", { name: "Toggle" })).toBeDisabled();
  });

  it("can have a toggle button with just an icon", () => {
    render(<ContextualMenu hasToggleIcon links={[]} />);
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(
        ".p-contextual-menu__toggle .p-contextual-menu__indicator"
      )
    ).toBeInTheDocument();
    const toggle = screen.getByRole("button", { name: Label.Toggle });
    expect(toggle.childNodes).toHaveLength(1);
  });

  it("makes the icon light if the toggle is positive or negative", () => {
    render(
      <ContextualMenu hasToggleIcon links={[]} toggleAppearance="positive" />
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(
        ".p-contextual-menu__toggle .p-contextual-menu__indicator"
      )
    ).toHaveClass("is-light");
  });

  it("can have a toggle button with an icon and text on the left", () => {
    render(
      <ContextualMenu
        hasToggleIcon
        links={[]}
        toggleLabel="Toggle"
        toggleLabelFirst
      />
    );
    const toggle = screen.getByRole("button", { name: "Toggle" });
    expect(toggle.childNodes[0].textContent).toBe("Toggle");
  });

  it("can have a toggle button with an icon and text on the right", () => {
    render(
      <ContextualMenu
        hasToggleIcon
        links={[]}
        toggleLabel="Toggle"
        toggleLabelFirst={false}
      />
    );
    const toggle = screen.getByRole("button", { name: "Toggle" });
    expect(toggle.childNodes[1].textContent).toBe("Toggle");
  });

  it("can have a toggle button with just text", () => {
    render(<ContextualMenu links={[]} toggleLabel="Toggle" />);
    expect(screen.getByRole("button", { name: "Toggle" })).toBeInTheDocument();
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-contextual-menu__indicator")
    ).not.toBeInTheDocument();
  });

  it("can not have a toggle button", () => {
    render(<ContextualMenu links={[]} />);
    expect(
      screen.queryByRole("button", { name: Label.Toggle })
    ).not.toBeInTheDocument();
  });

  it("can use the toggle button to display the menu", async () => {
    render(<ContextualMenu links={[]} toggleLabel="Toggle" />);
    expect(
      screen.queryByLabelText(DropdownLabel.Dropdown)
    ).not.toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Toggle", expanded: false })
    ).toBeInTheDocument();
    await userEvent.click(screen.getByRole("button", { name: "Toggle" }));
    expect(
      screen.getByRole("button", { name: "Toggle", expanded: true })
    ).toBeInTheDocument();
    expect(screen.getByLabelText(DropdownLabel.Dropdown)).toBeInTheDocument();
  });

  it("be visible without a toggle button", () => {
    render(<ContextualMenu links={[]} visible />);
    expect(screen.getByLabelText(DropdownLabel.Dropdown)).toBeInTheDocument();
  });

  it("can display links", () => {
    render(<ContextualMenu links={[{ children: "Link1" }]} visible />);
    expect(screen.getByRole("button", { name: "Link1" })).toBeInTheDocument();
  });

  it("can display links in groups", () => {
    render(<ContextualMenu links={[[{ children: "Link1" }]]} visible />);
    // eslint-disable-next-line testing-library/no-node-access
    const group = document.querySelector(
      ".p-contextual-menu__group"
    ) as HTMLElement;
    expect(group).toBeInTheDocument();
    expect(
      within(group).getByRole("button", { name: "Link1" })
    ).toBeInTheDocument();
  });

  it("can display a mix of links and groups", () => {
    render(
      <ContextualMenu
        links={[[{ children: "Link1" }], { children: "Link2" }]}
        visible
      />
    );
    // eslint-disable-next-line testing-library/no-node-access
    const group = document.querySelector(
      ".p-contextual-menu__group"
    ) as HTMLElement;
    expect(group).toBeInTheDocument();
    expect(
      within(group).getByRole("button", { name: "Link1" })
    ).toBeInTheDocument();
    expect(
      within(group).queryByRole("button", { name: "Link2" })
    ).not.toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Link2" })).toBeInTheDocument();
  });

  it("can supply content instead of links", () => {
    render(
      <ContextualMenu visible>
        <span data-testid="content">content</span>
      </ContextualMenu>
    );
    expect(screen.getByTestId("content")).toBeInTheDocument();
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-contextual-menu__link")
    ).not.toBeInTheDocument();
  });

  it("closes the menu when clicking on an item", async () => {
    render(
      <ContextualMenu
        links={[{ onClick: jest.fn(), children: "Link1" }]}
        toggleLabel="Toggle"
      />
    );
    // Open the menu:
    await userEvent.click(screen.getByRole("button", { name: "Toggle" }));
    expect(screen.getByLabelText(DropdownLabel.Dropdown)).toBeInTheDocument();
    // Click on an item:
    await userEvent.click(screen.getByRole("button", { name: "Link1" }));
    expect(
      screen.queryByLabelText(DropdownLabel.Dropdown)
    ).not.toBeInTheDocument();
  });

  it("can use a custom link component", () => {
    type LinkProps = {
      /** An address */
      to: string;
      children: React.ReactNode;
    };
    const Link = ({ children, to }: LinkProps) => <a href={to}>{children}</a>;
    render(
      <ContextualMenu
        links={[{ children: "Link1", element: Link, to: "http://example.com" }]}
        visible
      />
    );
    expect(screen.getByRole("link", { name: "Link1" })).toBeInTheDocument();
  });

  it("can pass props to the contextual menu", () => {
    render(<ContextualMenu links={[]} data-testid="extra-prop" />);
    expect(screen.getByTestId("extra-prop")).toBeInTheDocument();
    // Check that the prop was applied to the wrapping element:
    expect(screen.getByTestId("extra-prop")).toHaveClass("p-contextual-menu");
  });

  it("can pass props to the toggle button", () => {
    render(
      <ContextualMenu
        links={[]}
        toggleLabel="Toggle"
        toggleProps={{
          appearance: "negative",
          "data-testid": "extra-prop",
        }}
      />
    );
    expect(screen.getByTestId("extra-prop")).toBeInTheDocument();
    expect(screen.getByTestId("extra-prop")).toHaveClass("p-button--negative");
    // Check that the props were applied to the toggle element:
    expect(screen.getByTestId("extra-prop")).toHaveClass(
      "p-contextual-menu__toggle"
    );
  });

  it("can pass props to the contextual menu dropdown", async () => {
    render(
      <ContextualMenu
        links={[]}
        toggleLabel="Toggle"
        dropdownProps={{
          "data-testid": "extra-prop",
        }}
      />
    );
    // Open the menu.
    await userEvent.click(screen.getByRole("button", { name: "Toggle" }));
    expect(screen.getByTestId("extra-prop")).toBeInTheDocument();
    // Check that the props were applied to the dropdown element:
    expect(screen.getByTestId("extra-prop")).toHaveClass("p-contextual-menu");
  });

  it("allows passing a component as a toggleLabel", () => {
    render(
      <ContextualMenu
        links={[]}
        position="right"
        toggleLabel={<span>toggleLabel component text</span>}
      />
    );
    expect(screen.getByText("toggleLabel component text")).toBeInTheDocument();
  });
});
