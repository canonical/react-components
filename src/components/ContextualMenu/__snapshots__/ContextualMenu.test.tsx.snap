// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextualMenu  renders 1`] = `
<span
  class="p-contextual-menu"
  data-testid="menu"
  style="position: relative;"
>
  <span
    style="display: none;"
  />
</span>
`;
